pkg/controller/migration/convert/core.go

```diff
@@ -211,210 +443,372 @@ func checkNodeHostPathVolume(spec corev1.PodSpec, name, path string) error {
-			if annotationKeyRegexp.MatchString(key) {
-				delete(a, key)
-				break
-			}
+	// If TyphaAffinity is defined, convert it to a v1.Affinity and return it
+	if install.Spec.TyphaAffinity != nil && install.Spec.TyphaAffinity.NodeAffinity != nil {
+		return &corev1.Affinity{
+			NodeAffinity: &corev1.NodeAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution:  install.Spec.TyphaAffinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution,
+				PreferredDuringSchedulingIgnoredDuringExecution: install.Spec.TyphaAffinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution,
+			},
 		}
 	}
-
-	return a
+	return nil
 }
 
-// handleNodeSelectors is a migration handler which ensures that nodeSelectors are set as expected.
-// In general, setting custom nodeSelectors and nodeAffinity for components is not supported.
-// The exception to this is the calico-node nodeSelector, which is migrated into the
-// ControlPlaneNodeSelector field.
+// handleNodeSelectors is a migration handler which ensures that nodeSelectors and affinities are set as expected.
 func handleNodeSelectors(c *components, install *operatorv1.Installation) error {
-	// check calico-node nodeSelectors
+	// Handle calico-node.
 	if c.node.Spec.Template.Spec.Affinity != nil {
-		if !(install.Spec.KubernetesProvider == operatorv1.ProviderAKS && reflect.DeepEqual(c.node.Spec.Template.Spec.Affinity, &corev1.Affinity{
-			NodeAffinity: &corev1.NodeAffinity{
-				RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
-					NodeSelectorTerms: []corev1.NodeSelectorTerm{{
-						MatchExpressions: []corev1.NodeSelectorRequirement{{
-							Key:      "type",
-							Operator: corev1.NodeSelectorOpNotIn,
-							Values:   []string{"virtual-kubelet"},
-						}},
-					}},
-				},
-			},
-		})) && !(install.Spec.KubernetesProvider == operatorv1.ProviderEKS && reflect.DeepEqual(c.node.Spec.Template.Spec.Affinity, &corev1.Affinity{
-			NodeAffinity: &corev1.NodeAffinity{
-				RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
-					NodeSelectorTerms: []corev1.NodeSelectorTerm{{
-						MatchExpressions: []corev1.NodeSelectorRequirement{{
-							Key:      "eks.amazonaws.com/compute-type",
-							Operator: corev1.NodeSelectorOpNotIn,
-							Values:   []string{"fargate"},
-						}},
-					}},
-				},
-			},
-		})) {
-			return ErrIncompatibleCluster{
-				err:       "node affinity not supported for calico-node daemonset",
-				component: ComponentCalicoNode,
-				fix:       "remove the affinity",
+		if install.Spec.CalicoNodeDaemonSet == nil || install.Spec.CalicoNodeDaemonSet.GetAffinity() == nil {
+			// Affinity set on the component but not the installation so migrate it over.
+			helpers.EnsureCalicoNodePodSpecNotNil(install)
+			install.Spec.CalicoNodeDaemonSet.Spec.Template.Spec.Affinity = c.node.Spec.Template.Spec.Affinity
+		} else {
+			// Affinity is set on the component and the installation, verify that they match.
```

Comment:
Do we do this in other places of the migration? (Check that an installation setting matches what is being migrated.)
I'm thinking that if the user has specified an affinity that we should just use it. They've defined the setting so who knows if they want it to match what they previously had or not.